### DISCUSSION TITLE
When using the @Groups decorator in models - how can I make the TS type correct for each group?

### DISCUSSION THREAD

**Discussion by markusenglund** [TypedProject/ts-express-decorators#2304] [Q&A] [answered]
Hey!

Let's say I have a model that looks like this:

```typescript
export class Example {
    @Required()
    @RequiredGroups('!action.create')
    id?: string;

    @Required()
    content: string;

    constructor(options: any) {
        this.content = options.content;
        this.id = options.id
    }
}
```

The `id` property is not required when using the `action.create` group, but is required in all other cases. I want the code-base to be correctly typed, and _not_ have models incorrectly typed as having the `id` property during create when it won't be there at runtime. Is there a convenient way to generate two separate TS types for the two cases? Ideally without creating a separate class for each group? I tried to do the following:

```typescript
export class Example<T extends 'create' | 'get'> {
    @Required()
    @RequiredGroups('!action.create')
    id: T extends 'create' ? undefined : string;

    @Required()
    content: string;

    constructor(options: any) {
        this.content = options.content;
        this.platform = options.platform;
    }
}
``` 

However, the generated OpenAPI spec now says `id` has `"type": "object"`.

Any ideas for good solution?

**Comment by Romakita** [comment, marked as answer]
Hello @markusenglund 

When you use a complex interface like "T extends 'create' ? undefined : string;" , typescript doesn't set `String` as metadata but `Object`.

This is why Ts.ED isn't able to infer the correct type and this is why Ts.ED provide `@Property` decorator ;). The correct code is:

```ts
export class Example<T extends 'create' | 'get'> {
    @Property(String)
    @Required()
    @RequiredGroups('!action.create')
    id: T extends 'create' ? undefined : string;

    @Required()
    content: string;

    constructor(options: any) {
        this.content = options.content;
        this.platform = options.platform;
    }
}
```

See you
Romain
